src/module.c

```diff
@@ -677,29 +683,44 @@ int TSDB_createRule(RedisModuleCtx *ctx, RedisModuleString **argv, int argc) {
         return REDISMODULE_ERR;
     }
 
+    RedisModuleString *srcKeyName = argv[1];
     RedisModuleString *destKeyName = argv[2];
-    RedisModuleKey *destKey = RedisModule_OpenKey(ctx, destKeyName, REDISMODULE_READ);
-    if (RedisModule_KeyType(destKey) == REDISMODULE_KEYTYPE_EMPTY) {
-        return RedisModule_ReplyWithError(ctx, "TSDB: the destination key does not exist");
+    if(!RedisModule_StringCompare(srcKeyName, destKeyName)){
+    	return RedisModule_ReplyWithError(ctx, "TSDB: the source key and destination key should be different");
     }
-    if (RedisModule_ModuleTypeGetType(key) != SeriesType) {
-        return RedisModule_ReplyWithError(ctx, REDISMODULE_ERRORMSG_WRONGTYPE);
+
+    // First we verify the source is not a destination
+    Series *srcSeries;
+    int status = getSeries(ctx, srcKeyName, &srcSeries);
+    if(!status){
+    	return REDISMODULE_ERR;
+    }
+    if(srcSeries->srcKey){
+    	return RedisModule_ReplyWithError(ctx, "TSDB: the source key already has a source rule");
     }
 
-    Series *series = RedisModule_ModuleTypeGetValue(key);
-    if (SeriesHasRule(series, destKeyName)) {
-        return RedisModule_ReplyWithError(ctx, "TSDB: the destination key already has a rule");
+    // Second verify the destination doesn't have other rule
+    Series *destSeries;
+    status = getSeries(ctx, destKeyName, &destSeries);
+    if(!status){
+    	return REDISMODULE_ERR;
     }
+    srcKeyName = RedisModule_CreateStringFromString(ctx, srcKeyName);
+    if(!SeriesSetSrcRule(destSeries, srcKeyName)){
+    	return RedisModule_ReplyWithError(ctx, "TSDB: the destination key already has a rule");
+    }
+    RedisModule_RetainString(ctx, srcKeyName);
 
-    RedisModuleString *destKeyStr = RedisModule_CreateStringFromString(ctx, destKeyName);
-    if (SeriesAddRule(series, destKeyStr, aggType, bucketSize) == NULL) {
-    	return RedisModule_ReplyWithError(ctx, "ERROR creating rule");
+    // Last add the rule to source
+    destKeyName = RedisModule_CreateStringFromString(ctx, destKeyName);
+    if (SeriesAddRule(srcSeries, destKeyName, aggType, bucketSize) == NULL) {
+        RedisModule_ReplyWithSimpleString(ctx, "ERROR creating rule");
+        return REDISMODULE_ERR;
     }
-    RedisModule_RetainString(ctx, destKeyStr);
-    RedisModule_ReplyWithSimpleString(ctx, "OK");
+    RedisModule_RetainString(ctx, destKeyName);
```

Comment:
why not copy the string inside SeriesAddRule and retain it there?

Reply:
cause I noticed we don't pass context to this functions